source/libraries/ContractCompiler.ts

```diff
@@ -0,0 +1,141 @@
+
+        // Compile all contracts in the specified input directory
+        const compilerInputJson = await this.generateCompilerInput();
+        const compilerOutputJson = compileStandardWrapper(JSON.stringify(compilerInputJson));
+        const compilerOutput: CompilerOutput = JSON.parse(compilerOutputJson);
+        if (compilerOutput.errors) {
+            let errors = "";
+
+            for (let error of compilerOutput.errors) {
+                // FIXME: https://github.com/ethereum/solidity/issues/3273
+                if (error.message.includes("instruction is only available after the Metropolis hard fork")) continue;
+                errors += error.formattedMessage + "\n";
+            }
+
+            if (errors.length > 0) {
+              if (errors.match(/errors?:/i)) {
+                throw new Error("The following errors/warnings were returned by solc:\n\n" + errors);
+              } else {
+                console.warn("The following warnings were returned by solc\n\n" + errors);
+              }
+            }
+        }
+
+        // Create output directory (if it doesn't exist)
+        await asyncMkdirp(path.dirname(this.configuration.contractOutputPath));
+
+        // Output contract data to single file
+        const filteredCompilerOutput = this.filterCompilerOutput(compilerOutput);
+        await fs.writeFile(this.configuration.contractOutputPath, JSON.stringify(filteredCompilerOutput, null, '\t'));
+
+        // Output abi data to a single file
+        const abiOutput = this.generateAbiOutput(filteredCompilerOutput);
+        await fs.writeFile(this.configuration.abiOutputPath, JSON.stringify(abiOutput, null, '\t'));
+
+        return filteredCompilerOutput;
+    }
+
+    public async generateCompilerInput(): Promise<CompilerInput> {
+        const ignoreFile = function(file: string, stats: fs.Stats): boolean {
+            return file.indexOf("legacy_reputation") > -1 || (stats.isFile() && path.extname(file) !== ".sol");
+        }
+        const filePaths = await recursiveReadDir(this.configuration.contractSourceRoot, [ignoreFile]);
+        const filesPromises = filePaths.map(async filePath => (await readFile(filePath)).toString('utf8'));
+        const files = await Promise.all(filesPromises);
+
+        let inputJson: CompilerInput = {
+            language: "Solidity",
+            settings: {
+                optimizer: {
+                    enabled: true,
+                    runs: 500
+                },
+                outputSelection: {
+                    "*": {
+                        "*": [ "abi", "evm.bytecode.object" ]
+                    }
+                }
+            },
+            sources: {}
+        };
+        for (var file in files) {
+            const filePath = filePaths[file].replace(this.configuration.contractSourceRoot, "").replace(/\\/g, "/").replace("/src", "");
+            if (filePath.includes(".t.sol") || filePath.includes("test.sol")) continue;
```

Comment:
Avoid test files, at least for now. These cause other weird issues, too, like the 0-byte thing above I mentioned but didn't get a chance to look into